lib/charms/synapse/v0/matrix_auth.py

```diff
@@ -0,0 +1,512 @@
+        }
+        return dumped_data
+
+    @classmethod
+    def from_relation(cls, model: ops.Model, relation: ops.Relation) -> "MatrixAuthRequirerData":
+        """Get a MatrixAuthRequirerData from the relation data.
+
+        Args:
+            model: the Juju model.
+            relation: the relation.
+
+        Returns:
+            the relation data and the processed entries for it.
+
+        Raises:
+            ValueError: if the value is not parseable.
+        """
+        try:
+            app = cast(ops.Application, relation.app)
+            relation_data = relation.data[app]
+            registration_secret_id = (
+                (relation_data["registration_secret_id"])
+                if "registration_secret_id" in relation_data
+                else None
+            )
+            registration = MatrixAuthRequirerData.get_registration(model, registration_secret_id)
+            return MatrixAuthRequirerData(
+                registration=registration,
+            )
+
+        except json.JSONDecodeError as ex:
+            logger.warning("Invalid relation data %s", ex)
+            raise ValueError from ex
+
+
+#### Events ####
+class MatrixAuthRequestProcessed(ops.RelationEvent):
+    """MatrixAuth event emitted when a new request is processed."""
+
+    def get_matrix_auth_provider_relation_data(self) -> MatrixAuthProviderData:
+        """Get a MatrixAuthProviderData for the relation data.
+
+        Returns:
+            the MatrixAuthProviderData for the relation data.
+        """
+        return MatrixAuthProviderData.from_relation(self.model, self.relation)
+
+
+class MatrixAuthRequestReceived(ops.RelationEvent):
+    """MatrixAuth event emitted when a new request is made."""
+
+
+class MatrixAuthRequiresEvents(ops.CharmEvents):
+    """MatrixAuth requirer events.
+
+    This class defines the events that a MatrixAuth requirer can emit.
+
+    Attributes:
+        matrix_auth_request_processed: the MatrixAuthRequestProcessed.
+    """
+
+    matrix_auth_request_processed = ops.EventSource(MatrixAuthRequestProcessed)
+
+
+class MatrixAuthProvidesEvents(ops.CharmEvents):
+    """MatrixAuth provider events.
+
+    This class defines the events that a MatrixAuth provider can emit.
+
+    Attributes:
+        matrix_auth_request_received: the MatrixAuthRequestReceived.
+    """
+
+    matrix_auth_request_received = ops.EventSource(MatrixAuthRequestReceived)
+
+
+#### Requires and Provides ####
+class MatrixAuthRequires(ops.Object):
+    """Requirer side of the MatrixAuth requires relation.
+
+    Attributes:
+        on: events the provider can emit.
+    """
+
+    on = MatrixAuthRequiresEvents()
+
+    def __init__(self, charm: ops.CharmBase, relation_name: str = DEFAULT_RELATION_NAME) -> None:
+        """Construct.
+
+        Args:
+            charm: the provider charm.
+            relation_name: the relation name.
+        """
+        super().__init__(charm, relation_name)
+        self.charm = charm
```

Comment:
You might change this to get only the app, because `self.charm` is only used to get the app via `self.charm.model.app`

Reply:
It's even unecessary because you inherit model from `ops.Object`

Reply:
Closing this one since this library is hosted in the synapse repo.